cpmpy/solvers/choco.py

```diff
@@ -0,0 +1,585 @@
+                return self.chc_model.and_(self.solver_vars(cpm_expr.args))
+            elif cpm_expr.name == 'or':
+                return self.chc_model.or_(self.solver_vars(cpm_expr.args))
+            elif cpm_expr.name == '->':
+                cond, subexpr = cpm_expr.args
+                assert isinstance(cond, _BoolVarImpl), f"Implication constraint {cpm_expr} must have BoolVar as lhs"
+                lhs = self.solver_var(cond) # should always be boolvar due to only_bv_reifies
+                # right hand side
+                if isinstance(subexpr, _BoolVarImpl):
+                    # bv -> bv
+                    # PyChoco does not have "implies" constraint
+                    return self.chc_model.or_([~lhs, self.solver_var(subexpr)])
+                else:
+                    # bv -> boolexpr
+                    # the `reify_rewrite()` transformation ensures that only reifiable rhs remain here
+                    if subexpr.name == 'not':
+                        bv = self._get_constraint(subexpr.args[0]).reify()
+                        return self.chc_model.or_([~lhs, ~bv])
+                    else:
+                        bv = self._get_constraint(cpm_expr.args[1]).reify()
+                        return self.chc_model.or_([~lhs, bv])
+            else:
+                raise NotImplementedError("Not a known supported Choco Operator '{}' {}".format(
+                    cpm_expr.name, cpm_expr))
+
+        # Comparisons: both numeric and boolean ones
+        # numexpr `comp` bvar|const
+        # Choco accepts only int32, not int64
+        elif isinstance(cpm_expr, Comparison):
+            for i in range(len(cpm_expr.args)):
+                if isinstance(cpm_expr.args[i], np.integer):
+                    cpm_expr.args[i] = int(cpm_expr.args[i])
+            lhs, rhs = cpm_expr.args
+
+            if lhs.is_bool() and rhs.is_bool(): #boolean equality -- Reification
+                if isinstance(rhs, _NumVarImpl):
+                    return self.chc_model.all_equal(self.solver_vars([lhs, rhs]))
+                else:
+                    bv = self._get_constraint(rhs).reify()
+                    return self.chc_model.all_equal([self.solver_var(lhs), bv])
+            elif isinstance(lhs, _NumVarImpl) or (isinstance(lhs, Operator) and (
+                    lhs.name == 'sum' or lhs.name == 'wsum' or lhs.name == "sub")):
+                # a BoundedLinearExpression LHS, special case, like in objective
+                chc_numexpr = self._make_numexpr(cpm_expr)
+                return chc_numexpr
+            elif cpm_expr.name == '==':
+                chcrhs = self.solver_var(rhs)
+                # NumExpr == IV, supported by Choco (thanks to `only_numexpr_equality()` transformation)
+                if lhs.name == 'min':
+                    if isinstance(rhs, int):     # Choco does not accept an int in rhs
+                        chcrhs = self.to_var(rhs)  # convert to "variable"
```

Comment:
Choco also does not accept ints in any of the arguments of min, max, abs,.. something like Max(iv,0) also needs the 0 to become a variable.